Use `SampleRate` type consistently

Holzhaus opened the pull request:
This replaces all remaining occurrences of integer sample rates that I found via `git grep` and uses the semantic `mixxx::audio::SampleRate` type instead.

Holzhaus committed:
refactor(graphiceqeffect): Use `constexpr` instead of `#define`

Holzhaus committed:
refactor(engine): Use `SampleRate` type instead of `int`/`unsigned int`

Holzhaus committed:
refactor(beatmaptest): Use `SampleRate` type instead of `int`

Holzhaus committed:
refactor(track/key*): Use `SampleRate` type instead of `int`

Holzhaus committed:
refactor(vinulcontrolxwax): Use `SampleRate` type instead of `int`

Holzhaus committed:
refactor(soundio): Use `SampleRate` type instead of `unsigned int`

Holzhaus committed:
refactor(cachingreader): Use `SampleRate` type instead of `int`

Holzhaus committed:
refactor(beatgridtest): Use `SampleRate` type instead of `int`

Holzhaus committed:
refactor(bpmcontrol_test): Use `SampleRate` type instead of `int`

Holzhaus committed:
refactor(cuecontrol_test): Use `SampleRate` type instead of `int`

Holzhaus committed:
refactor(searchqueryparsertest): Remove unnecessary `sampleRate` arg

Holzhaus committed:
refactor(enginevumeter): Use `SampleRate` type instead of `int`

daschuer commented on src/soundio/soundmanagerconfig.cpp:
```diff
@@ -409,9 +410,8 @@ unsigned int SoundManagerConfig::getFramesPerBuffer() const {
-    double sampleRate = m_sampleRate; // need this to avoid int division
     // first, get to the framesPerBuffer value corresponding to latency index 1
-    for (; framesPerBuffer / sampleRate * 1000 < 1.0; framesPerBuffer *= 2) {
+    for (; framesPerBuffer / m_sampleRate.toDouble() * 1000 < 1.0; framesPerBuffer *= 2) {
```
does this need to be double math? (Not tested)
```suggestion
    for (; framesPerBuffer / (m_sampleRate * 1000) < 1; framesPerBuffer *= 2) {
```
The compiler should be able to move m_sampleRate * 1000 in front of the loop.

daschuer replied:
That looks good.

daschuer reviewed (changes requested):
Thank you. I left some comments.

Swiftb0y commented on src/effects/backends/builtin/filtereffect.cpp:
```diff
@@ -128,8 +128,8 @@ void FilterEffect::processChannel(
         }
-        pState->m_pLowFilter->setFrequencyCorners(1, lpf, clampedQ);
-        pState->m_pHighFilter->setFrequencyCorners(1, hpf, clampedQ);
+        pState->m_pLowFilter->setFrequencyCorners(mixxx::audio::SampleRate(1), lpf, clampedQ);
```
but the parameter is named samplerate. What should it be named instead if its not used as a samplerate?

daschuer replied:
It can be refactored into using our mixxx::audio::SampleRate that is always in samples/s. 
Currently it looks it sets the filter to a 1/s sample rate.

Holzhaus replied:
I'm unsure what you want me to do TBH.

daschuer replied:
I am not sure if we have a common understanding about the issue here. Do we? 

I propose to revert the Fidlib related changes here and merge the rest of this PR quickly. This way we can consider the best solution for this issue in a smaller PR < 66 files .

Holzhaus replied:
I checked in the fidlib code and the `sampleRate` parameter is actually passed as `double rate` parameter of the `fid_design()` method. In the error handling code, rate is used as the second argument of the error message `Frequency of %gHz out of range with sampling rate of %gHz`. So it seems indeed the `sampling rate`. No idea why we pass just `1` though.

Should I revert the changes nonetheless?

daschuer replied:
So I think we have two options. 
1. Postpone the Fidlib related actions 
2. Use here the real 1/s base sample rate values via our semantic type and adjust the other parameters accordingly.

Holzhaus replied:
Undoing all the fidlib-related changes turned out tedious, so I opted for 2. Please have a look at https://github.com/mixxxdj/mixxx/pull/11904/commits/99e1088b6c7566ec15ab23ff9153b197b75af3e6.

Holzhaus committed:
chore(enginevumeter): Remove `i` prefix from `m_iSamplesCalculated`

Holzhaus commented on src/preferences/dialog/dlgprefsound.cpp:
```diff
@@ -435,7 +436,8 @@ void DlgPrefSound::loadSettings(const SoundManagerConfig &config) {
     }
-    int sampleRateIndex = sampleRateComboBox->findData(m_config.getSampleRate());
+    int sampleRateIndex = sampleRateComboBox->findData(
+            QVariant::fromValue(m_config.getSampleRate()));
```
What do you mean? `mixxx::audio::SampleRate` is declared as `QT_METATYPE` and can therefore be directly stored in a `QVariant`.

daschuer replied:
Nice, than the toUint() call below is wong.

Holzhaus replied:
Fixed in f1bdd02e1f52f819bcc077ab8caf7b3ba513e522.

Holzhaus commented on src/soundio/soundmanager.cpp:
```diff
@@ -559,8 +559,8 @@ SoundDeviceStatus SoundManager::setConfig(const SoundManagerConfig& config) {
     // Do this first so vinyl control gets the right samplerate -- Owen W.
-    m_pConfig->set(ConfigKey("[Soundcard]","Samplerate"),
-                   ConfigValue(static_cast<int>(m_config.getSampleRate())));
+    m_pConfig->set(ConfigKey("[Soundcard]", "Samplerate"),
+            ConfigValue(static_cast<int>(m_config.getSampleRate().value())));
```
```
/home/jan/Projects/mixxx/src/soundio/soundmanager.cpp: In member function ‘SoundDeviceStatus SoundManager::setConfig(const SoundManagerConfig&)’:
/home/jan/Projects/mixxx/src/soundio/soundmanager.cpp:563:57: error: call of overloaded ‘ConfigValue(mixxx::audio::SampleRate::value_t)’ is ambiguous
  563 |             ConfigValue(m_config.getSampleRate().value()));
      |                                                         ^
In file included from /home/jan/Projects/mixxx/src/preferences/usersettings.h:6,
                 from /home/jan/Projects/mixxx/src/controllers/midi/midimessage.h:10,
                 from /home/jan/Projects/mixxx/src/control/controlbehavior.h:6,
                 from /home/jan/Projects/mixxx/src/control/control.h:9,
                 from /home/jan/Projects/mixxx/src/control/pollingcontrolproxy.h:6,
                 from /home/jan/Projects/mixxx/src/soundio/soundmanager.h:11,
                 from /home/jan/Projects/mixxx/src/soundio/soundmanager.cpp:1:
/home/jan/Projects/mixxx/src/preferences/configobject.h:83:14: note: candidate: ‘ConfigValue::ConfigValue(double)’
   83 |     explicit ConfigValue(double value);
      |              ^~~~~~~~~~~
/home/jan/Projects/mixxx/src/preferences/configobject.h:82:14: note: candidate: ‘ConfigValue::ConfigValue(int)’
   82 |     explicit ConfigValue(int value);
      |              ^~~~~~~~~~~
/home/jan/Projects/mixxx/src/preferences/configobject.h:73:7: note: candidate: ‘ConfigValue::ConfigValue(const ConfigValue&)’
   73 | class ConfigValue {
      |       ^~~~~~~~~~~
```

Swiftb0y replied:
okay, Ignore for now. I'll look into it some other time.

Swiftb0y replied:
I've looked into it. The problem is trivial on the surface (uint32_t can't be promoted to int nor double, thus it would be converted and neither conversion ranks better than the other). The problem gets harder when I looked into fixing it. `configobject.h` is unnecessarily complex with lots of duplicated code due to a lack of separation of concerns (`ConfigValue` vs `ConfigObject::setValue`, which type/function is supposed to do the conversion?). I may do some refactoring in the future.

Holzhaus committed:
chore(soundmanagerconfig): Simplify calculation of `framesPerBuffer`

Holzhaus committed:
chore(dlgprefsound): Use `const auto&` instead of `auto&` for loop

Holzhaus commented on src/engine/enginevumeter.cpp:
```diff
@@ -53,14 +54,14 @@ EngineVuMeter::~EngineVuMeter()
     m_fRMSvolumeSumR += fVolSumR;
 
-    m_iSamplesCalculated += iBufferSize / 2;
+    m_iSamplesCalculated += static_cast<unsigned int>(iBufferSize / 2);
```
Yup.

Holzhaus committed:
refactor(soundio): Use `SampleRate` type instead of `double`

Holzhaus committed:
refactor(soundsourcemad): Use `SampleRate` type instead of `unsigned int`

Holzhaus committed:
fix(dlgprefsound): Directly read `SampleRate` type from `QVariant`

Swiftb0y reviewed (approved):
My comments have been resolved, waiting for LGTM from @daschuer

Holzhaus committed:
refactor(filtereffect): Pass actual sample rate and remove normalization

daschuer commented on src/soundio/sounddevicenotfound.h:
```diff
@@ -35,7 +35,7 @@ class SoundDeviceNotFound : public SoundDevice {
 
-    unsigned int getDefaultSampleRate() const override {
-        return 44100;
+    mixxx::audio::SampleRate getDefaultSampleRate() const override {
+        return SoundDevice::kFallbackSampleRate;
```
We have a naming issue with kFallbackSampleRate: 

SoundManagerConfig::kFallbackSampleRate = mixxx::audio::SampleRate(48000);
constexpr mixxx::audio::SampleRate SoundDevice::kFallbackSampleRate = mixxx::audio::SampleRate(44100);

Than we have:
constexpr auto kDefaultSampleRate = mixxx::audio::SampleRate(44100);
And various definitions in global namespaces. 

I think the term kFallbackSampleRate has been invented for the 48000 Hz that all soundcards likely support. 
The 44100 is Mixxx's default sample rate that matches CD rips and is the most common sample rate when the decision has been made. 

I suggest to use the term kFallbackSampleRate for 48000 Hz and kDefaultSampleRate for 44100 or maybe even better: 
kMixxxDefaultSampleRate .. 

Than you may check which additional constants can be removed.

Holzhaus replied:
I renamed it.

> Than you may check which additional constants can be removed.

I don't see what this has to do with this PR to be honest. As you already mentioned yourself, this PR is quite big and I don't get why we keep expanding the scope of PRs even further. I think its totally fine to fix one thing without also fixing some only tangentially related other thing.

daschuer reviewed (changes requested):
Removing the normalization form the filter seems to work. Thank you. 
I have left two additional comments.

Holzhaus committed:
chore(networkoutputstreamworker): Remove redundant `SampleRate` initializer

Holzhaus committed:
refactor(modplug): Use `SampleRate`/`ChannelCount` instead of `int`

Holzhaus committed:
chore(soundio): Rename device fallback rate to `kMixxxDefaultSampleRate`

Holzhaus committed:
chore(effects): Use actual engine sample rate instead of constant

We know the actual sample rate from the `EngineParameters` object, so
there is no need to hardcode some constant.

daschuer commented on src/soundio/soundmanagerconfig.cpp:
```diff
@@ -10,8 +10,10 @@
 const QString SoundManagerConfig::kEmptyComboBox = QStringLiteral("---");
-// Sample Rate even the cheap sound Devices will support most likely
-const unsigned int SoundManagerConfig::kFallbackSampleRate = 48000;
+constexpr mixxx::audio::SampleRate SoundManagerConfig::kMixxxDefaultSampleRate =
+        mixxx::audio::SampleRate(44100);
+constexpr mixxx::audio::SampleRate SoundManagerConfig::kFallbackSampleRate =
+        mixxx::audio::SampleRate(48000);
```
```suggestion
const mixxx::audio::SampleRate SoundManagerConfig::kMixxxDefaultSampleRate =
        mixxx::audio::SampleRate(44100);
const mixxx::audio::SampleRate SoundManagerConfig::kFallbackSampleRate =
        mixxx::audio::SampleRate(48000);
```
This produces a linker error with ldd.
It should be anyway better to match the definition in the header.

Swiftb0y replied:
make both constexpr, move to header instead? Or would you consider that to leak implementation details?

Holzhaus replied:
https://github.com/mixxxdj/mixxx/pull/11904/commits/987fd3a461a8133cd073e5285b157f524a025ca3 ok?

daschuer reviewed (changes requested):
One remaining issue. The rest looks and works good. Thank you.

Holzhaus committed:
chore(soundio): Move sample rate constants to `SoundManagerConfig` header

This is supposed to fix `ldd` linker errors that may occur with the
previous code.

See https://github.com/mixxxdj/mixxx/pull/11904#discussion_r1318891197
for details.

daschuer reviewed (approved):
LGTM, Thank you

Swiftb0y commented:
Thank you
